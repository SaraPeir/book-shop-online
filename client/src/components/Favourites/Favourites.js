import React, {useContext} from 'react';
import './Favourites.scss';
import { Container, Row, Col } from 'reactstrap';
import ShoppingCard from '../ShoppingCard';

const Favourites = (props) => {
  const getData = (cachedData) => {
    if(cachedData) return cachedData
    return [] 

  }

  const booksData = getData(props.cachedData);

  const favourites = booksData.filter(book => book.isFavourite);

  // para controlar que la cache se actualiza instantaneamente gracias a refetchQueries
  console.log('favourites length', favourites.length);

  if(!favourites.length) return 'Non hai selezionato nessun elemento come favorito'; 
      return (
        <Container>
          <p className="main-style">{props.title}</p>
          <Row xs="1" sm="2" md="2" lg="3" xl="4">
              {favourites.map((book, id) =>  
                book.isFavourite && 
                <Col key={id}>
                  <ShoppingCard 
                    cardStyle={'card-style'}
                    updateBooks={() => console.log('updateBooks')} 
                    isFavourite={book.isFavourite} 
                    id={book.id} 
                    image = {book.image} 
                    author={book.author} 
                    title={book.title} 
                    price={book.price} 
                    isForFavourites= {true}
                  />
                </Col>
              )}
            </Row>
        </Container>
      );
  }

  export default Favourites;