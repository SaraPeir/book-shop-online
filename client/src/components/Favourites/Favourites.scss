.favourites-container {
    position: relative;
    width: 80%;
    margin-left: auto;
    margin-right: auto;
    border: 1px solid red;
}

.card-style {
    width: 280px;
    margin: 10px auto;
}