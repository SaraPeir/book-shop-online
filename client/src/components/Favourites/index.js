import Tripboard from './Favourites';

export default Tripboard;