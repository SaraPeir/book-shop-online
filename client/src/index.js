import React from 'react';
import ReactDOM from 'react-dom';
import './index.scss';
import App from './App';
import 'bootstrap/dist/css/bootstrap.min.css';
import { ApolloClient, HttpLink, InMemoryCache, ApolloProvider  } from '@apollo/client';

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: new HttpLink({
    uri: 'http://localhost:4000/graphql'
  })
});

console.log('client', client);

ReactDOM.render(
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>,
  document.getElementById('root')
);
