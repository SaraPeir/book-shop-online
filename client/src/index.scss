@import "variables.scss";

.main-style {
    color: $primary-color;
}