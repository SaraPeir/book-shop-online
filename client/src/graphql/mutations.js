import { gql } from '@apollo/client';

export const UPDATE_BOOKS = gql`
    mutation updateBook($id: ID, $isFavourite: Boolean) {
        updateBook(input: {
        id: $id,
        isFavourite: $isFavourite
        }) {
            title
            description
            id
            author
            price
            score
            count
            pages
            image
            isFavourite
            type
        }
    }
`