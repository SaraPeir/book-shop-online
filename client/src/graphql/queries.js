import { gql } from '@apollo/client';

export const GET_BOOKS  = gql`
query getBooks {
    getBooks {
      title
      description
      id
      author
      price
      score
      count
      pages
      image
      isFavourite
      type
    }
  }
` 