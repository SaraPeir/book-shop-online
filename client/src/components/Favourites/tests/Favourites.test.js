import React from 'react';
import Favourites from '../Favourites';
import { MockedProvider } from '@apollo/client/testing';
import {UPDATE_BOOKS} from '../../../graphql/mutations';
import {GET_BOOKS} from '../../../graphql/queries.js';
// import { act } from 'react-dom/test-utils';

import { render, 
   fireEvent, 
   waitFor,
    screen } from '@testing-library/react'

    // npm test -- --coverage --watchAll=false para el report
    const mockedFn = jest.fn();

    const mocks = [
        {
          request: {
            query: UPDATE_BOOKS,
            variables: {
              id: "5eb83d4b7e177a8a54277468",
              isFavourite: false
            },
          },
          result: {
            data: {
                updateBook: {
                id: "5eb83d4b7e177a8a54277468",
                isFavourite: true
              },
            },
          }
       
        },
        {
            request: {
              query: GET_BOOKS, // necessary to mock refetching after mutation
              variables: {},
            },
            result: {
              data: {
                  getBook: [
                    {
                        author: "AUTHOR_1",
                        count: 3,
                        description: "BOOK1_DESCIPTION.",
                        id: "5eb83d4b7e177a8a54277468",
                        image: "IMG_1",
                        isFavourite: false,
                        pages: 312,
                        price: "€ 18.00",
                        score: 4,
                        title: "TITLE_1",
                        type: "TYPE_1"
                    },
                    {
                        author: "AUTHOR_2",
                        count: 3,
                        description: "BOOK3_DESCIPTION.",
                        id: "5eb83d4b7e177a8a54277468",
                        image: "IMG_2",
                        isFavourite: true,
                        pages: 312,
                        price: "€ 18.00",
                        score: 4,
                        title: "TITLE_2",
                        type: "TYPE_2"
                    },
                    {
                        author: "AUTHOR_3",
                        count: 3,
                        description: "BOOK3_DESCIPTION.",
                        id: "5eb83d4b7e177a8a54277468",
                        image: "IMG_3",
                        isFavourite: true,
                        pages: 312,
                        price: "€ 18.00",
                        score: 4,
                        title: "TITLE_3",
                        type: "TYPE_3"
                    }
                  ],
              },
            }
          },
      ]

const data = {
    cachedBook: [
        {
            author: "AUTHOR_1",
            count: 3,
            description: "BOOK1_DESCIPTION.",
            id: "5eb83d4b7e177a8a54277468",
            image: "IMG_1",
            isFavourite: true,
            pages: 312,
            price: "€ 18.00",
            score: 4,
            title: "TITLE_1",
            type: "TYPE_1"
        },
        {
            author: "AUTHOR_2",
            count: 3,
            description: "BOOK3_DESCIPTION.",
            id: "5eb83d4b7e177a8a54277468",
            image: "IMG_2",
            isFavourite: true,
            pages: 312,
            price: "€ 18.00",
            score: 4,
            title: "TITLE_2",
            type: "TYPE_2"
        },
        {
            author: "AUTHOR_3",
            count: 3,
            description: "BOOK3_DESCIPTION.",
            id: "5eb83d4b7e177a8a54277468",
            image: "IMG_3",
            isFavourite: true,
            pages: 312,
            price: "€ 18.00",
            score: 4,
            title: "TITLE_3",
            type: "TYPE_3"
        }  
    ]
}

let props = {
    title: 'I tuoi preferiti',
    cachedData: data.cachedBook
}

it('Should render correctly', () => {
    const {container} = render(
        <MockedProvider mocks={mocks}>
            <Favourites {...props} updateBooks={mockedFn} />
        </MockedProvider>
      );
  
      expect(container).toMatchSnapshot()
})

it('Should not render cards if there are no data', () => {
    const noDataProps = {
        title: 'I tuoi preferiti'
    }
    const {container} = render(
        <MockedProvider mocks={mocks}>
            <Favourites {...noDataProps} updateBooks={mockedFn} />
        </MockedProvider>
      );
  
      expect(container).toMatchSnapshot()
})

it('Should execute a function when you click at close button', () => {
    let button;
    const {container, getAllByTestId, rerender} = render(
        <MockedProvider mocks={mocks}>
            <Favourites {...props} updateBooks={mockedFn} />
        </MockedProvider>
      );
      button = getAllByTestId('close-button')[0];
      expect(button).toBeInTheDocument()
  
      //   // solucion desde aquí: https://jkettmann.com/testing-apollo-how-to-test-if-a-mutation-was-called-with-mockedprovider/
    //   const addBookMutationMock = mocks[0].result;
    //   addBookMutationMock()
    //   expect(addBookMutationMock).toHaveBeenCalled()
        fireEvent.click(button);

    setTimeout(() => expect(mockedFn).toHaveBeenCalled(), 100);
})