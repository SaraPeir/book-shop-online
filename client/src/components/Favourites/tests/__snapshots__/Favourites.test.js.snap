// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should not render cards if there are no data 1`] = `
<div>
  Non hai selezionato nessun elemento come favorito
</div>
`;

exports[`Should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <p
      class="main-style"
    >
      I tuoi preferiti
    </p>
    <div
      class="row row-cols-1 row-cols-sm-2 row-cols-md-2 row-cols-lg-3 row-cols-xl-4"
    >
      <div
        class="col"
      >
        <div
          class="card-style card"
        >
          <div
            class="close-button-icon-container"
            data-testid="close-button"
          >
            X
          </div>
          <div
            class="img-container"
          >
            <img
              alt="Card image cap"
              class="card-img"
              src="IMG_1"
            />
          </div>
          <div
            class="card-body"
          >
            <div
              class="card-title"
            >
              TITLE_1
            </div>
            <div
              class="card-subtitle"
            >
              AUTHOR_1
            </div>
            <div
              class="price-container"
            >
              <p
                class="card-text"
              >
                € 18.00
              </p>
            </div>
            <div
              class="btn-container"
            >
              <button
                class="btn btn-secondary"
              >
                Dettagli
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col"
      >
        <div
          class="card-style card"
        >
          <div
            class="close-button-icon-container"
            data-testid="close-button"
          >
            X
          </div>
          <div
            class="img-container"
          >
            <img
              alt="Card image cap"
              class="card-img"
              src="IMG_2"
            />
          </div>
          <div
            class="card-body"
          >
            <div
              class="card-title"
            >
              TITLE_2
            </div>
            <div
              class="card-subtitle"
            >
              AUTHOR_2
            </div>
            <div
              class="price-container"
            >
              <p
                class="card-text"
              >
                € 18.00
              </p>
            </div>
            <div
              class="btn-container"
            >
              <button
                class="btn btn-secondary"
              >
                Dettagli
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col"
      >
        <div
          class="card-style card"
        >
          <div
            class="close-button-icon-container"
            data-testid="close-button"
          >
            X
          </div>
          <div
            class="img-container"
          >
            <img
              alt="Card image cap"
              class="card-img"
              src="IMG_3"
            />
          </div>
          <div
            class="card-body"
          >
            <div
              class="card-title"
            >
              TITLE_3
            </div>
            <div
              class="card-subtitle"
            >
              AUTHOR_3
            </div>
            <div
              class="price-container"
            >
              <p
                class="card-text"
              >
                € 18.00
              </p>
            </div>
            <div
              class="btn-container"
            >
              <button
                class="btn btn-secondary"
              >
                Dettagli
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
